docs/source/extending_pywr/extending_pywr_nodes.rst

```diff
@@ -0,0 +1,116 @@
+.. _extending-pywr-nodes:
+
+Extending Pywr with custom Nodes
+--------------------------------
+
+Nodes and subclasses thereof provide the basic network structure in Pywr. There are several different types
+of node available. Two major categories exist: flow nodes and storage nodes. Flow nodes are the typical nodes
+that represent rivers, pipes and other features from, through or to which a resource can flow. These nodes are
+typically characterised by minimum and maximum flow rates. Storage nodes provide the ability to store resource
+from one time-step to another, and are characterised by minimum, maximum and initial volumes.
+
+There are three fundamental types of node in Pywr:
+
+* ``Input`` nodes add water to the system
+* ``Output`` nodes remove water from the system
+* ``Link`` nodes do not add or remove water from the system
+
+There is a fourth node type, ``Storage``, which can be considered fundamental because there are special rules for
+it's behaviour in the linear programme used to solve the water balance:
+
+* ``Storage`` nodes can carry water from one timestep to the next
+
+All other node types in Pywr are subclasses of these base types. For example, the ``pywr.nodes.Catchment`` node type
+is a special case of ``Input`` where the ``min_flow`` and ``max_flow`` properties are equal.
+
+The most common way to create a new node type is using a compound node. A compound node contains one or more existing
+nodes and is used to manage common or more complex arrangements of the basic node types. An example of a compound node
+is the ``PiecewiseLink``. It is composed of a link (``OUT_1``) which receives water from upstream and an link
+(``IN_1``) which conveys water downstream, connected by a set of links in parallel (``LNK_1`` ... ``LNK_N``) each with
+a different ``max_flow`` and ``cost``, illustrated below::
+
+                               /-->-- LNK_1 -->--\
+    UPSTREAM -->-- OUT_1 -->--|--->--  ...  -->---|-->-- IN_1 -->-- DOWNSTREAM
+                               \-->-- LNK_N -->--/
+
+Let's look at an example to create a new node type that represents a leaky pipe. To remove water from the system we
+need to use an output node (``LEAK``), with two links representing the boundaries of the compound node (``INFLOW`` and
+``OUTFLOW``)::
+
+    UPSTREAM -->-- INFLOW -->-- OUTFLOW -->-- DOWNSTREAM
```

Comment:
This raises the same question as #642 . Because `self` is not part of the of network (i.e. `allow_isolated=True`) there is weird behaviour if you try to set `self.max_flow` for example. It will not error, but it will also not apply any constraint to this compound node structure.

Should it be like this?
```
UPSTREAM --> SELF -->-- INFLOW -->-- OUTFLOW -->-- DOWNSTREAM
                                 |
                                 \------>--  LEAK
```

Reply:
In this case you could probably drop the INFLOW node.

Reply:
Due to the way `Node.connect` and `Node.iter_slots` are implemented it's actually not possible to do this without manipulating the model's graph directly.

```
        # Connect the internal nodes
        # self.connect(self.outflow)
        # self.connect(self.leak)
        model.graph.add_edge(self, self.outflow)
        model.graph.add_edge(self, self.leak)
```

Reply:
Why does the `self.connect(self.outflow)` not work?